### DISCUSSION TITLE
Having some issues with creating object type for generic interface

### DISCUSSION THREAD

**Discussion by jnssnmrcs** [hayes/pothos#876] [Q&A] [answered]
Hi!

I'm trying to create the following function for creating a object type for generic input, but I cannot get it to work:

```ts
interface PageInfo {
  hasPreviousPage: boolean;
  hasNextPage: boolean;
  startCursor: string;
  endCursor: string;
}

interface Edge<T> {
  node: T;
  cursor: string;
}

interface Connection<T> {
  edges: Edge<T>[];
  pageInfo: PageInfo;
}

const PageInfo = builder.objectRef<PageInfo>('PageInfo').implement({
  fields: (t) => ({
    hasPreviousPage: t.exposeBoolean('hasPreviousPage'),
    hasNextPage: t.exposeBoolean('hasNextPage'),
    startCursor: t.field({ type: 'Cuid', nullable: true, resolve: (i) => i.startCursor }),
    endCursor: t.field({ type: 'Cuid', nullable: true, resolve: (i) => i.endCursor }),
  }),
});

export const buildConnectionType = <T>(name: string, nodeType: ObjectRef<T>) => {
  const Edge = builder.objectRef<Edge<T>>(`${name}Edge`).implement({
    fields: (t) => ({
      node: t.field({
        resolve: (parent) => parent.node,
        type: nodeType,
      }),
      cursor: t.field({
        resolve: (parent) => parent.cursor,
        type: 'Cuid',
      }),
    }),
  });

  const Connection = builder.objectRef<Connection<T>>(`${name}Connection`).implement({
    fields: (t) => ({
      edges: t.field({
        resolve: (parent) => parent.edges,
        type: [Edge],
      }),
      pageInfo: t.field({
        resolve: (parent) => parent.pageInfo,
        type: PageInfo,
      }),
    }),
  });

  return Connection;
};

const User = builder.objectType('User', {
  fields: (t) => ({
    username: t.field({
      type: 'String',
      resolve: (u) => u.username,
    }),
    password: t.field({
      type: 'String',
      resolve: (u) => u.password,
    }),
  }),
});

const UserConnection = buildConnectionType('User', User);
```

but I'm getting the following typescript error for the parent.node resolver:

![image](https://user-images.githubusercontent.com/6576541/231573641-e12138b8-ac26-42fc-83e9-2fd052c41be2.png)

```
Type 'T' is not assignable to type '[T] extends [readonly (infer Item)[] | null | undefined] ? ListResolveValue<T, Item, T> : MaybePromise<T>'.
```

I Don't really understand what I'm doing wrong so I was hoping someone could help me. Thanks :)

**Comment by jnssnmrcs** [comment]
Just to add: I just tested and it works without any issues if I just @ts-ignore the compiler error...

**Reply by hayes** [reply, marked as answer]
This is a side effect of some complicated types intended to make resolver errors more clear.  It doesn't work well with genetics because the generic might be an array.  It should be safe to ignore

**Comment by hayes** [comment]
Curious if there is a reason you are not using the relay plugin to implement this pattern?

**Reply by jnssnmrcs** [reply]
Actually I totally missed that it existed, but I'm also just trying to learn pothos so I guess this was a good exercise :D Thanks for quick answers!

**Reply by jnssnmrcs** [reply]
Sorry if this is the wrong place to ask but I was reading up on the relay pagination spec and it says this about order of edges:

> The ordering of edges should be the same when using first/after as when using last/before, all other arguments being equal. It should not be reversed when using last/before.

And I was also checking out the code example from the relay plugin where it does this:

```ts
builder.queryField('posts', (t) =>
  t.connection({
    type: Post,
    resolve: (_, args) =>
      resolveCursorConnection(
        {
          args,
          toCursor: (post) => post.createdAt.toISOString(),
        },
        // Manually defining the arg type here is required
        // so that typescript can correctly infer the return value
        ({ before, after, limit, inverted }: ResolveCursorConnectionArgs) =>
          prisma.post.findMany({
            take: limit,
            where: {
              createdAt: {
                lt: before,
                gt: after,
              },
            },
            orderBy: {
              createdAt: inverted ? 'desc' : 'asc',
            },
          }),
      ),
  }),
);
```

Here it's actually reversing the order of the items it returns depending on the last/first arguments passed, isn't that breaking the relay spec?

**Reply by jnssnmrcs** [reply]
Nevermind I checked out the source code and found it's actually reversing back the order of items before returning them.

**Reply by hayes** [reply]
Good question!  The helper will reverse the list.  The point here is to simplify the API so you don't have to worry about first/last etc.  When inverted is true, you just invert your ordering when querying so you still can use cursors/limits, and then the helper will reverse it so that it complies with what relay spec expects

**Reply by jnssnmrcs** [reply]
Thats actually really smart, thanks for clarifying

**Reply by hayes** [reply]
To be a little more explicit.  It's really the limit that benefits from having the querying inverted.  I haven't used this helper much myself, so let me know if this seems like the wrong way to implement the API

**Reply by jnssnmrcs** [reply]
As soon as I want to let the api user also order the items, I can't use that helper anymore right? because it relies on items being ordered by the cursor to do the lt and gt where-query. I managed to create my own version that also allows for ordering, but only allows first/after OR last/before, by using prisma cursor functionality.

**Reply by hayes** [reply]
I don't think there is any reason it wouldn't work with user ordering, you still have access to all the args (I found a few connections in our codebase that do this).

That being said, if you are using Prisma, the Prisma plugin will work much better for creating connections and makes things way simpler

**Reply by jnssnmrcs** [reply]
I think I'm tripping over my own thoughts here, and I also totally missed that the prisma plugin also supported generating connections....thanks again!

**Reply by jnssnmrcs** [reply]
Could you point me towards the places in the code base you do ordering together with the resolveCursorConnection as it is used above (by using where-clause with lt and gt)  because I can't get it to work..

**Reply by hayes** [reply]
I don't think I have anything public for that specific use case (I meant the API I work on for work)

Basically what you need to do is use the before/after as normal (on whatever column the ordering is happening on). Then for the orderBy apply it as given by the user unless inverted is true, in which case you would invert the ordering the user requested.


This inversion is required to ensure that limits apply from the correct end of the range.


For Prisma I'd highly recommend the Prisma plugin.  You can also just use the connection helper from the Prisma plugin.  

The cursor helper is for cases where the db doesn't support cursors natively.  Prisma actually has a cursor pagination API which might be cleaner to use (this is what the Prisma plugin does). But wouldn't work well with this helper.


This is what the Prisma plugin uses: https://github.com/hayes/pothos/blob/main/packages/plugin-prisma/src/util/cursors.ts#L319

**Reply by jnssnmrcs** [reply]
Thanks for explaining, i will check out the prisma plugin but i just wanted to understand how it would work for my own sanity :D really appreciate it!
